lib/chef/provider/package/msu.rb

```diff
@@ -0,0 +1,146 @@
+#
+# Author:: Nimisha Sharad (<nimisha.sharad@msystechnologies.com>)
+# Copyright:: Copyright 2015-2016, Chef Software, Inc.
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# msu_package leverages cab_package
+# The contents of msu file are extracted, which contains one or more cab files.
+# The extracted cab files are installed using Chef::Resource::Package::CabPackage
+# Reference: https://support.microsoft.com/en-in/kb/934307
+require "chef/provider/package"
+require "chef/resource/msu_package"
+require "chef/mixin/shell_out"
+require "chef/provider/package/cab"
+require "chef/util/path_helper"
+require "chef/mixin/uris"
+require "chef/mixin/checksum"
+
+class Chef
+  class Provider
+    class Package
+      class Msu < Chef::Provider::Package
+        use_inline_resources
+        include Chef::Mixin::ShellOut
+        include Chef::Mixin::Uris
+        include Chef::Mixin::Checksum
+
+        provides :msu_package, os: "windows"
+
+        def load_current_resource
+          @current_resource = Chef::Resource::MsuPackage.new(new_resource.name)
+
+          # download file if source is a url
+          msu_file = uri_scheme?(new_resource.source) ? download_source_file : @new_resource.source
+
+          # temp directory where the contents of msu file get extracted
+          @temp_directory = Dir.mktmpdir("chef")
+          extract_msu_contents(msu_file, @temp_directory)
+          @cab_files = read_cab_files_from_xml(@temp_directory)
+
+          unless @cab_files.empty?
+            # set the package version
+            # Here is a catch: How to set package version in case of multiple cab files?
+            cab_resource = @new_resource
+            cab_resource.source = @cab_files.first
+            cab_pkg = Chef::Provider::Package::Cab.new(cab_resource, nil)
+
+            new_resource.version(cab_pkg.package_version)
```

Comment:
> @lamont-granquist , there isn't any command to get the package version for msu package. 

I think the question here is, why are we setting `new_resource.version`, not why are we using the version from the CAB instead of the version form the MSU.

Normally `new_resource.version` is the version requested by the user by setting `version` on the resource. If the MSU version is meaningless, we should be able to pass the CAB version in `new_resource.version`. If the action is set to install, with a version set, and the underlying CAB provider cannot install that exact version, we should fail.